### PR TITLE
DOC-440 K8s guides for 'Developer Guides'

### PR DESCRIPTION
**Pull request by bmansheim**
Merged from #1008 for style review

### CODE REVIEW

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 1]
```diff
@@ -0,0 +1,501 @@
+---
+Title: Configuring LDAP Authentication
+description: This task describes how to configure LDAP-based authentication for a cluster on Kubernetes.
```
Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:3]:
```suggestion
description: How to configure LDAP-based authentication for a Redis Enterprise cluster on Kubernetes.
```

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+weight: 90
+alwaysopen: false
+categories: ["Platforms"]
+aliases:
+---
+
+
+Throughout this tutorial we will assume there is a single namespace called ‘db'.
```
Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:11]:
```suggestion
This tutorial describes how to configure LDAP-based authentication for a Redis Enterprise cluster on Kubernetes. Throughout this tutorial, we will assume that there is a single namespace called ‘db'.
```

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+
+## LDAP Overview
```
Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:13]:
```suggestion
## LDAP overview
```

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+Then you change the LDAP configuration with a JSON version of the ldap configuration.
+
+```
+cat << EOF > tmp.json
+{"saslauthd_ldap_conf": "ldap_servers: ldap://ldap-openldap.bdb.svc:389\nldap_search_base: ou=users,dc=example,dc=org\nldap_filter: (uid=%u)\nldap_bind_dn: cn=admin,dc=example,dc=org\nldap_password: admin\n"}
+EOF
+curl -f -k -u "demo@redislabs.com:xxx" -X PUT -d @tmp.json -H "Content-Type: application/json" https://localhost:9443/v1/cluster
+```
+
+The key value for “saslauthd_ldap_conf” is the configuration file verbatim as it would be in the saslauthd configuration file. As such, it can be relatively tricky to get the syntax right and generating this programmatically is helpful. A little python program (make_request.py) can help:
```
Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:181]: Please ensure that the names of processes like saslauthd are between back ticks: `saslauthd`

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+```
+
+And so we can do:
+
+```
+python make_request.py saslauthd_ldap_conf ldap.conf > tmp.json
+curl -f -k -u "demo@redislabs.com:xxx" -X PUT -d @tmp.json -H "Content-Type: application/json" https://localhost:9443/v1/cluster
+```
+
+### Automating LDAP Configuration
```
Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:200]: Please change all titles to sentence case

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+               - name: config
+                 mountPath: "/config"
+                 readOnly: true
+         volumes:
+           - name: config
+             configMap:
+               name: config-job
+   ```
+
+   In the above job, the cluster name is “ldap”. You will have to change this to the name of your cluster in all three of the CLUSTER_NAME, CLUSTER_USER, and CLUSTER_PASSWORD environment variables.
```
Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:253]: Can you choose a more realistic cluster name here? It may confuse the user to call the cluster "ladp"

Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:253]: Please put env variables in back ticks

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+
+This job waits for the cluster to be in the “Running” state. You can submit this job at the same time as the Redis Enterprise CR. It will wait for a period of time for the cluster to spin up and then change the configuration. This time period is tunable.
```
Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:255]: CR?

Comment (banker) [reply content/platforms/kubernetes/tasks/ldap-on-k8s.md:255]: Please avoid starting sentences with "It", as the antecedent is often unclear.

Comment (banker) [reply content/platforms/kubernetes/tasks/ldap-on-k8s.md:255]: Suggest removing "This time period is tunable". Seems irrelevant here.

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+
+You'll need to create the docker image (i.e., in the example as "you/job-config-ldap:1")
```
Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:257]:
```suggestion
You'll need to create the Docker image (i.e., in the example as "you/job-config-ldap:1")
```

Comment (banker) [reply content/platforms/kubernetes/tasks/ldap-on-k8s.md:257]: Can you be more specific about what Docker image the reader needs to create? This is the first mention of a Docker image here.

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 9]
```diff
@@ -0,0 +1,501 @@
...
+and make it available to your K8s cluster.
+
+You can build this docker image yourself with the following:
```
Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:260]:
```suggestion
You can build this Docker image yourself with the following:
```

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 10]
```diff
@@ -0,0 +1,501 @@
...
+     configMap:
+       name: ldap-certs
+  redisEnterpriseVolumeMounts:
+   - name: ldap-volume
+     mountPath: /opt/ldap
+```
+
+The result is we can use the file path “/opt/ldap/ca.crt” for the “ldap_tls_cert_file” parameter.
+
+You can use the usual techniques to map a set of files into a directory for the “ldap_tls_cert_dir” parameter if you have multiple CA certificates:
```
Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:446]: Can you link "the usual techniques" to some official documentation?

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md (hunk 2 above)
Comment (banker) [reply content/platforms/kubernetes/tasks/ldap-on-k8s.md:11]: @alexmilowski Can you summarize the steps of the tutorial up front? There's a lot of information here, and a little roadmap would be useful.

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 11]
```diff
@@ -0,0 +1,501 @@
...
+ldap_tls_cert: /opt/ldap/client.crt
+ldap_tls_key: /opt/ldap/client.key
+```
+
+## Known Issues
+
+1. The saslauthd daemon is not restarted when the configuration changes. If you have never configured LDAP, it will still work. The configuration file does not exist until you do and so it will get read.
+Unfortunately, if you change the configuration and have previously tried to authenticate with an external user, the configuration will not get read again until the saslauthd daemon is restarted. There is currently no way to do that via K8s other than to restart the pods because K8s administrators don’t have access to root in the Redis Enterprise container.
+
+   You can proceed with caution and restart the pods by:
```
Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:493]:
```suggestion
   You can proceed with caution and restart the pods by running:
```

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 12]
```diff
@@ -0,0 +1,501 @@
...
+
+   ```
+   kubectl rollout restart statefulset name-of-cluster
+   ```
+
+   where "name-of-cluster" is the name of the Redis Enterprise cluster in your CR.
+
+1. For cluster deployed LDAP servers the suffix “cluster.local” will not resolve properly. You can remove the cluster.local suffix for LDAP deployments in the same K8s cluster.
```
Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:501]:
```suggestion
1. For cluster-deployed LDAP servers, the suffix “cluster.local” will not resolve properly. You can remove the "cluster.local" suffix for LDAP deployments in the same K8s cluster.
```

Comment (banker) [reply content/platforms/kubernetes/tasks/ldap-on-k8s.md:501]: You can remove or you should remove?

File: content/platforms/kubernetes/tasks/replica-of-k8s.md [hunk 13]
```diff
@@ -0,0 +1,253 @@
+---
+Title: Creating replica databases on Kubernetes
+description: This task describes how to create and automate database replicas via the database controller
```
Comment (banker) [inline content/platforms/kubernetes/tasks/replica-of-k8s.md:3]:
```suggestion
description: How to create and automate database replicas using the database controller
```

File: content/platforms/kubernetes/tasks/replica-of-k8s.md [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+weight: 42
+alwaysopen: false
+categories: ["Platforms"]
+aliases:
+---
+
+Configuring a replica of database can be accomplished by creating an item in
```
Comment (banker) [inline content/platforms/kubernetes/tasks/replica-of-k8s.md:10]: Please use the active voice: https://developers.google.com/style/voice

File: content/platforms/kubernetes/tasks/using-kustomnize.md [hunk 15]
```diff
@@ -0,0 +1,167 @@
+---
+Title: Using kustomize for deployment
+description: This tasks describes how to use the kustomize tool with the Redis Enterprise operator on Kubernetes
```
Comment (banker) [inline content/platforms/kubernetes/tasks/using-kustomnize.md:3]:
```suggestion
description: How to use the kustomize tool with the Redis Enterprise operator on Kubernetes
```

**Review by banker — CHANGES_REQUESTED** [review]

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 16]
```diff
@@ -0,0 +1,501 @@
...
+
+Throughout this tutorial we will assume there is a single namespace called ‘db'.
+
+## LDAP Overview
+
+At minimum, you’ll need the following information about your LDAP server:
+
+ * The base name for where to find users (e.g., ou=users,dc=example,dc=org)
+ * The filter to identify a particular user (e.g., uid=%u)
+ * The bind DN for the account that can search (e.g., cn=admin,dc=example,dc=org)
```
[the remark this answers is not in the document]
Comment (amiramm) [reply content/platforms/kubernetes/tasks/ldap-on-k8s.md:19]: DN stands for Distinguished Name - but anyone that touches LDAP uses this nomenclature and rarely refers to it by the full name, from my experience.

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 17]
```diff
@@ -0,0 +1,501 @@
...
+ * The password for the bind DN
+ * The server host and port.
+ * Whether you are using SSL
+
+In this tutorial, we will use a test server deployed on the same K8s cluster.
+
+## Setting up an LDAP server
+
+If you don’t already have an LDAP server, you can easily deploy one in the same K8s
+cluster for testing using a [helm chart for OpenLDAP](https://hub.helm.sh/charts/stable/openldap):
```
Comment (amiramm) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:29]:
```suggestion
cluster for testing using a [helm chart for OpenLDAP](https://github.com/helm/charts/tree/master/stable/openldap):
```

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 18]
```diff
@@ -0,0 +1,501 @@
...
+## Setting up an LDAP Server
+
+If you don’t already have an LDAP server, you can easily deploy one in the same K8s
+cluster for testing using a [helm chart for OpenLDAP](https://hub.helm.sh/charts/stable/openldap):
+
+1. Create a ldap-values.yaml file that contains:
+
+   ```
+   env:
+     LDAP_ORGANISATION: "ACME Inc."
```
[the remark this answers is not in the document]
Comment (amiramm) [reply content/platforms/kubernetes/tasks/ldap-on-k8s.md:35]: @banker This is a parameter name the helm chart expects, not a spelling choice.

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 19]
```diff
@@ -0,0 +1,501 @@
...
+   configPassword: config
+   ```
+
+1. Deploy OpenLDAP:
+
+   ```
+   helm install ldap -f ldap-values.yaml stable/openldap
+   ```
+
+1. Forward the pod to your localhost so you can configure the users. The ‘xxx’ should be replace with the actual pod hash:
```
[the remark this answers is not in the document]
Comment (amiramm) [reply content/platforms/kubernetes/tasks/ldap-on-k8s.md:47]: No. https://en.wikipedia.org/wiki/Localhost

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 20]
```diff
@@ -0,0 +1,501 @@
...
+   userPassword: {SSHA}5IcQ2zo5wkCCohXHGWteDMBDbJElbChP
+   ```
+
+   The password for the user is “tester”. If you’d like to change it, you can with:
+
+   ```
+   ldappasswd  -h localhost -p 3889 -s newpassword -W -D "cn=admin,dc=example,dc=org" -x "uid=tester,ou=users,dc=example,dc=org"
+   ```
+
+1. Create the OU and user:
```
[the remark this answers is not in the document]
Comment (amiramm) [reply content/platforms/kubernetes/tasks/ldap-on-k8s.md:82]: I don't think we should call out OU.

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 21]
```diff
@@ -0,0 +1,501 @@
...
+    "role":"admin",
+    "email_alerts":false,
+    "auth_method":"external"
+}
+EOF
+curl -v -k -u "demo@redislabs.com:xxx" -X POST -d @add-user.json -H "Content-Type: application/json" https://localhost:9443/v1/users
+```
+
+### Deploying a Redis Enterprise Cluster
+A Redis Enterprise cluster requires no special setup other to be configured with LDAP as the configuration is outside of the scope of the Redis Enterprise CR.
```
[the remark this answers is not in the document]
Comment (amiramm) [reply content/platforms/kubernetes/tasks/ldap-on-k8s.md:119]: Custom Resource. https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 22]
```diff
@@ -0,0 +1,501 @@
...
+      memory: 3Gi
+    requests:
+      cpu: 1
+      memory: 3Gi
+```
+
+### LDAP Configuration
+
+A new Redis Enterprise cluster does not have any LDAP configuration but is ready
+to accept them. The saslauthd daemon is already running and configured to use
```
[the remark this answers is not in the document]
Comment (amiramm) [reply content/platforms/kubernetes/tasks/ldap-on-k8s.md:142]: I think you're changing the meaning here...

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 23]
```diff
@@ -0,0 +1,501 @@
...
+LDAP as a mechanism. All you need to do is to provide the cluster
```
[the remark this answers is not in the document]
Comment (amiramm) [reply content/platforms/kubernetes/tasks/ldap-on-k8s.md:143]: also, here

**Review by amiramm — APPROVED** [review]

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 24]
```diff
@@ -0,0 +1,501 @@
...
+    "role":"admin",
+    "email_alerts":false,
+    "auth_method":"external"
+}
+EOF
+curl -v -k -u "demo@redislabs.com:xxx" -X POST -d @add-user.json -H "Content-Type: application/json" https://localhost:9443/v1/users
+```
+
+### Deploying a Redis Enterprise Cluster
+A Redis Enterprise cluster requires no special setup other to be configured with LDAP as the configuration is outside of the scope of the Redis Enterprise CR.
```
Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:119]:
```suggestion
A Redis Enterprise cluster requires no special setup other to be configured with LDAP as the configuration is outside of the scope of the Redis Enterprise custom resource.
```

File: content/platforms/kubernetes/tasks/ldap-on-k8s.md [hunk 25]
```diff
@@ -0,0 +1,501 @@
...
+                 readOnly: true
+         volumes:
+           - name: config
+             configMap:
+               name: config-job
+   ```
+
+   In the above job, the cluster name is “ldap”. You will have to change this to the name of your cluster in all three of the CLUSTER_NAME, CLUSTER_USER, and CLUSTER_PASSWORD environment variables.
+
+This job waits for the cluster to be in the “Running” state. You can submit this job at the same time as the Redis Enterprise CR. It will wait for a period of time for the cluster to spin up and then change the configuration. This time period is tunable.
```
Comment (banker) [inline content/platforms/kubernetes/tasks/ldap-on-k8s.md:255]:
```suggestion
This job waits for the cluster to be in the “Running” state. You can submit this job at the same time as the Redis Enterprise custom resource. It will wait for a period of time for the cluster to spin up and then change the configuration. This time period is tunable.
```

**Review by banker — APPROVED** [review]

File: content/platforms/kubernetes/tasks/replica-of-k8s.md [hunk 26]
```diff
@@ -0,0 +1,253 @@
+---
+Title: Creating replica databases on Kubernetes
+description: How to create and automate database replicas using the database controller
+weight: 42
+alwaysopen: false
+categories: ["Platforms"]
+aliases:
+---
+
+You can configure a replica of database by creating an item in
```
Comment (banker) [inline content/platforms/kubernetes/tasks/replica-of-k8s.md:10]:
```suggestion
You can configure a replica of a database by creating an item in
```

File: content/platforms/kubernetes/tasks/using-kustomnize.md [hunk 27]
```diff
@@ -0,0 +1,167 @@
...
+
+patchesStrategicMerge:
+ - cluster.yaml
+EOF
+```
+
+At this, we can test our “dev” customization by running the folliwing:
+
+```
+kubectl kustomize dev
```
Comment (banker) [inline content/platforms/kubernetes/tasks/using-kustomnize.md:118]: This confused me, since you earlier say to run it using `kubectl apply -k`

**Comment by banker** [comment]
@amiramm or @alexmilowski: Can you take one last look at my changes here? If you agree, I'll merge.

File: content/platforms/kubernetes/tasks/using-kustomnize.md (hunk 27 above)
Comment (alexmilowski) [reply content/platforms/kubernetes/tasks/using-kustomnize.md:118]: `kubectl apply -k` applys the configuration to the cluster and `kubectl kustomize dev` just runs the kustomize process locally and shows you the output.

Comment (alexmilowski) [reply content/platforms/kubernetes/tasks/using-kustomnize.md:118]: I fixed that sentence now.

**Comment by alexmilowski** [comment]
I've reviewed the changes. They all look good. I fixed one typo/clarification in the first use of the kustomize command as per the comment above. This is ready to be merged if you agree with that change.
